protocols/request-response/src/handler.rs

```diff
@@ -381,22 +397,55 @@ where
         cx: &mut Context<'_>,
     ) -> Poll<ConnectionHandlerEvent<Protocol<TCodec::Protocol>, (), Self::ToBehaviour, Self::Error>>
     {
+        // Drain pending events that were produced before poll.
+        // E.g. `Event::IncomingRequest` produced by `on_fully_negotiated_inbound`.
+        //
+        // NOTE: This is needed because if `read_request` fails before reaching a
+        // `.await` point, the incoming request will never register and `debug_assert`
+        // in `InboundStreamFailed` will panic.
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(event));
+        } else if self.pending_events.capacity() > EMPTY_QUEUE_SHRINK_THRESHOLD {
+            self.pending_events.shrink_to_fit();
+        }
+
         loop {
             match self.worker_streams.poll_unpin(cx) {
                 Poll::Ready((_, Ok(Ok(event)))) => {
-                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(event))
+                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(event));
                 }
-                Poll::Ready((id, Ok(Err(e)))) => {
+                Poll::Ready(((id, direction), Ok(Err(e)))) => {
                     log::debug!("Stream for request {id} failed: {e}");
+
+                    let event = match direction {
+                        Direction::Inbound => Event::InboundStreamFailed {
+                            request_id: id,
+                            error: e,
+                        },
+                        Direction::Outbound => Event::OutboundStreamFailed {
+                            request_id: id,
+                            error: e,
+                        },
+                    };
+
+                    // TODO: How should we handle errors produced after ConnectionClose event?
+                    // `ConnectionClose` will generate its own error. But only one of the two
+                    // should be forwarded to the upper layer.
+                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(event));
```

Comment:
@thomaseizinger I don't know what is the proper way to fix this. Also I don't think it is easily testable. This is the scenario:

* `Handler::on_fully_negotiated_inbound` will add in `worker_streams` the task for the request and emmit `IncomingRequest`
* `Behaviour::on_connection_handler_event` handles `IncomingRequest` and adds the `request_id` in `pending_inbound_responses`
* The other end closes the connection
* `Behaviour::on_connection_closed` is called, connection is removed, and `InboundFailure` is generated for all `pending_inbound_responses`
* `Handler::poll` runs and polls the `worker_stream`, the task of the connection fails and `Poll::Ready(Err(_))` is returned
* `Handler::poll` produces `InboundStreamFailed` event
* `Behavior::on_connection_handler_event` handles it but panics on `debug_assert` of the `InboundStreamFailed` arm. In case of release, two errors are generated for a single request.

I see three possible ways to fix this:

1. `Behaviour::on_connection_closed` should also cancel tasks from `worker_streams`.
2. `Handler::poll` should not generate errors for already closed connections. `Handler` needs to be informed by `Behaviour`.
3. Instead of having assertion in `Behaviour::on_connection_handler_event`, we can generate an error if `removed` is `true`, otherwise we ignore it.

What do you prefer?

Reply:
After some thinking, I will implement the 1. I will add cancellation functionality in futures-bounded.

Reply:
As per open maintainers call, my hypothesis is not correct. `Swarm` drops the `Connection` when `ConnectionClosed` is emitted so the task itself will not generate any errors. In this case `Behaviour` will generate `InboundFailure` to inform the upper layer.